### DISCUSSION TITLE
Memory requirement issues for grain growth simulations

### DISCUSSION THREAD

**Discussion by sahu07** [idaholab/moose#19437] [Q&A General] [answered]
Dear MOOSE team,

I am trying to resolve an out-of-memory problem in my simulations. For which I would like to know the purpose of "split grains" and its impact on memory requirement. **Can you please talk about the memory requirement for "split grains"? Do more split grains require more memory?**

I did two groups of grain growth simulations, namely "grainsize25" and "grainsize35". Both the groups of simulations have identical input scripts except for the Dream3D input microstructure (INL file). "grainsize25" means grain size is 25 grids and "grainsize35" means grain size is 35. All other input parameters are the same ( 128X128x64 mesh; num_steps=1; 30 order parameters; 10 MPI processors with 4 threads each; No mesh adaptivity; )
But the memory utilized were drastically different: ( 5 trials each)
**grainsize25**: (57.78 GB; 57.72 GB; 57.74 GB; 57.74 GB; 57.74 GB)
**grainsize35**: (150.97 GB; 106.49 GB; 110.60 GB; 147.49 GB; 75.41 GB)

Why "grainsize35" simulations have higher memory? I thought it is due to "split grains" but I don't know the internal workings of split grain. **How to reduce the memory requirement for "grainsize35"?** Similar problem occurs in "grainsize50" and "grainsize60", where the memory requirement exceeds 1 TB. **Could you help me to resolve this issue?**

Thank you!
Regards,
Sarita

**Comment by ykvishal** [comment, marked as answer]
I don't know the reason for  "split grains" issue, but I want to confirm one thing.  Are you using [EBSDReader](https://mooseframework.inl.gov/source/userobjects/EBSDReader.html)? My colleague and I found that EBSDReader takes a lot of memory per core. Can you provide your input files (.i and .txt)? I can quickly check your memory usage with [MemoryUsage](https://mooseframework.inl.gov/source/postprocessors/MemoryUsage.html) and [VectorMemoryUsage](https://mooseframework.inl.gov/source/vectorpostprocessors/VectorMemoryUsage.html).

**Reply by sahu07** [reply]
Hi Vishal, 

Yes, I am using EBSDReader. 

Here are the input and output files.

[grainsize25_a.i.txt](https://github.com/idaholab/moose/files/7573218/grainsize25_a.i.txt)
[grainsize25_a.out.txt](https://github.com/idaholab/moose/files/7573219/grainsize25_a.out.txt)
[grainsize35_a.i.txt](https://github.com/idaholab/moose/files/7573220/grainsize35_a.i.txt)
[grainsize35_a.out.txt](https://github.com/idaholab/moose/files/7573221/grainsize35_a.out.txt)

Thanks, 
Sarita

**Reply by ykvishal** [reply]
For grainsize25 case, the number of grains is 149 at t=0 and t=1. For grainsize35 case, the number of grains at  t=0 and  t=1 are 47 and 115 , respectively. **ie the number of grains are increasing!!!!!** Without looking at image, I can only guess that you have a large number of tiny regions (ie approximately equal to one mesh element), and it may be the reason for increasing number of grains ( and for the split grain warning). 

Also, do you need 30 order parameters? If you can reduce it (apply trial and error approach) then your memory requirement will also decrease. There is another issue (other than huge memory requirement) with EBSDReader - the GrainTracker algorithm does not work efficiently at t=0 time-step. My colleague and I  are going to post these issues (huge memory requirement and problem with GrainTracker at time-step=0) related to EBSDReader here.
